Column reorder feature

joshbax189 opened the pull request:
Allows setting an order for columns with a defcustom per mode. You can also hide columns if you choose, and set column widths.
Added this to containers, images, networks and volumes.

**Implementation notes:**
I added a default order alist that matches the data order of the format string in `docker-X-entries`. 
This alist is also used as the default value for the defcustom `docker-X-column-order`.

Then `docker-X-parse` computes the reordering by matching the default order with the user's specified order and permutes the data to match.

The tablist headers are computed from `docker-X-column-order` too.

Both of these steps are implement with common functions in `docker-utils`.

**Specific feedback**
- My elisp is still pretty rough, in particular I feel that the four new `docker-util` functions could be improved.
- Should the contents of `docker-X-column-order` match the format of `tabulated-list-format`? This would remove the need for `docker-utils-column-order-list-format`, but would complicate the presentation of the customize widget. (That's why I didn't implement it like that in the first place).

Silex commented:
@joshbax189: Can you rebase this PR? it looks awesome :+1:

joshbax189 commented:
Cheers! Rebased it just now.

Fair enough about the TODOs. Would you prefer I make PRs a single commit? I.e. I squash my own commits rather than you squash merging them into the main branch.

Silex commented:
I can squash it when merging, but I looked a bit and I have several inputs:

- It looks like "Hidden" does not work, I tried on images by setting the ID column to Hidden and there was errors.
- `"This should match the column order used in the format string in docker-image-entries."`, I wonder if we could not avoid such potential bugs and use `docker-image-default-column-order` when building `  (let* ((fmt "[{{json .Repository}},{{json .Tag}},{{json .ID}},{{json .CreatedAt}},{{json .Size}}]")`
  - pro: there is just one source for the column order / name.
  - con: we have to use the docker name (e.g `CreatedAt` instead of `Created`).
  - maybe: create an entity like `Column(DockerName, DisplayName, Width)` if you see what I mean.
- The defcustom for the column order is not very self-understanding ("Choice", "Integer"), it'd be better of it displayed "Name" or "Width".

Silex commented:
Basically something like:

``` elisp
(defconst docker-image-default-column-order
  '((:id "Repository" :name "Repository" :width 30)
    (:id "Tag" :name "Tag"  :width 20)
    (:id "ID" :name "Id" :width 16)
    (:id "CreatedAt" :name "Created" :width 23)
    (:id "Size" :name "Size" :width 10))
  "This should match the column order used in the format string in docker-image-entries.")
```

joshbax189 commented:
Ah yep, I had thought something like `(:name "Id" :width 16 :template "{{ json .ID}}" )` would remove the ambiguity in how the `:id` props were treated, and gives you the flexibility to write transforms into the templates. 

Do you think it's worth exposing templates in the customization menu too?

joshbax189 commented:
RE: the "hidden" bug, there was a typo in `docker-image` that I fixed in the rebase, it should work now.

Silex commented:
> Ah yep, I had thought something like `(:name "Id" :width 16 :template "{{ json .ID}}" )` would remove the ambiguity in how the `:id` props were treated, and gives you the flexibility to write transforms into the templates.

I like this. Do you want to give it a try?

About "Hidden" I think removing it in favor of plain add/remove of column(s) is more logical. It's also future proof if more columns are available the user can edit them in.

See for example:

``` elisp
(defcustom docker-image-default-columns
  '(("Id" 16 "{{ json .ID }}")
    ("Created" 24 "{{ json .CreatedAt }}"))
  "Fixme"
  :group 'docker-image
  :type '(repeat (list :tag "Column" (string :tag "Name") (integer :tag "Width") (string :tag "Template"))))

(defcustom docker-image-default-columns2
  '((:name "Id" :width 16 :template "{{ json .ID }}")
    (:name "Created" :width 24 :template "{{ json .CreatedAt }}"))
  "Fixme"
  :group 'docker-image
  :type '(repeat (plist :tag "Column" string integer string)))
```

`docker-image-default-columns` has the `customize-variable` looking the way I want but the lisp is not as verbose and `docker-image-default-columns2` has the list looking like I want but the `customize-variable` is a bit stupid.

joshbax189 commented:
Working on it.
> It's also future proof if more columns are available the user can edit them in.

Amusingly, that was going to be my next PR!

It looks like I can use `:get` and `:set` on `defcustom` to use the plist data format with the nicer `:type` metadata.

joshbax189 commented:
Ok so columns are now described like
```elisp
(defcustom docker-container-column-spec
  '((:name "Id" :width 16 :template "{{json .ID}}" :sort nil :format nil)
    (:name "Image" :width 15 :template "{{json .Image}}" :sort nil :format nil)
    (:name "Command" :width 30 :template "{{json .Command}}" :sort nil :format nil)
    (:name "Created" :width 23 :template "{{json .CreatedAt}}" :sort nil :format (lambda (x) (format-time-string "%F %T" (date-to-time x))))
    (:name "Status" :width 20 :template "{{json .Status}}" :sort nil :format (lambda (x) (propertize x 'font-lock-face (docker-container-status-face x))))
    (:name "Ports" :width 10 :template "{{json .Ports}}" :sort nil :format nil)
    (:name "Names" :width 10 :template "{{json .Names}}" :sort nil :format nil))
```
I added getters and setters that format the columns with appropriate widgets
![docker-reorder](https://user-images.githubusercontent.com/30708316/134460214-7890da34-88bb-4110-b7c0-3075a27d9d14.gif)

you can reorder them like so, and hide just by deleting the entry.

There are also custom format and sort functions.

In order to not break the Id extract function (used when marking lines in tablist mode) the Id field is specified separately:
```elisp
(defconst docker-container-id-template
  "{{ json .Names }}"
  "This Go template extracts the container id which will be passed to transient commands.")
```
and the format string is roughly `[id-template, ...other-templates]`. The parse function splits the id-tempate off, so it isn't printed in the tablist view.

There is a dependency between the `default-sort-key` variable and the `column-spec` one. Ideally `default-sort-key` would become another property in the column specifications, but removing a defcustom would probably mean a major version bump, so I left it for now. Instead I added a validation function to `default-sort-key`. (It doesn't catch the case where you rename the default sort column though).

One other minor quirk I noticed: due to the getter/setters on `column-spec` the menu always shows "edited" after the column spec submenu is opened.

Silex commented:
Awesome! I need to look at it and I'll give feedback.

Silex commented:
I have almost nothing to say, the code looks clean and much simpler/generic than before.

I think some names could be improved, e.g `docker-container-column-spec` is maybe better as `docker-container-columns-definition` ? or simply `docker-containers-columns` ?

What do you think?

Silex commented:
BTW, are you interested in being a maintainer? I'm a new father and my emacs time drastically reduced so I think you would help a lot keeping this package more active :wink:

joshbax189 commented:
> BTW, are you interested in being a maintainer? I'm a new father and my emacs time drastically reduced so I think you would help a lot keeping this package more active wink

Congratulations! I'd be glad to help out, just tag me if you need a hand with reviews, issues, or planned features.

Silex commented:
🎉 thanks!

By the way, in the customize-variable for me "DEL" just deletes, it does not "save for later insertion". Any idea why? I use Emacs 27.2

Silex commented:
I added you as a collaborator. Feel free to create branches directly here but go through a PR and review with me before merging to master 😉

Silex commented:
By the way I now noticed several things:

- You forgot `docker-machine.el`, was that intentional?
- Before `docker-container-default-sort-key` was a choice option, now it's a string but you validate its value. I think we can still easily get the choice option out of `docker-container-columns`, no?

Silex commented:
@joshbax189: friendly ping 😉

joshbax189 commented:
"Forgot" not quite, I've just never used it. That said, it should be a straightforward task to copy changes over. I did notice that docker-machine is now deprecated. So I guess we'd push over generic features without doing anything specific for docker-machines?

For the sort-key option: I'd be in favour of moving it inside the columns plist as a boolean. That way it would avoid a bi-directional dependency between the two.

joshbax189 commented:
> By the way, in the customize-variable for me "DEL" just deletes, it does not "save for later insertion". Any idea why? I use Emacs 27.2

Ah Damn. I couldn't find a specific mention in the news file, but my emacs version is 28.05, and my 27.2 version matches your observation...

joshbax189 commented:
I'm definitely in favour of removing it too! Feel free to merge or close #165 as you see fit.
